WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift

```diff
@@ -0,0 +1,97 @@
+import Foundation
+
+final class CustomFieldsListViewModel: ObservableObject {
+    private let originalCustomFields: [CustomFieldViewModel]
+
+    var shouldShowErrorState: Bool {
+        savingError != nil
+    }
+
+    @Published private(set) var savingError: Error?
+    @Published private(set) var combinedList: [CustomFieldUI] = []
+
+    @Published private var editedFields: [CustomFieldUI] = []
+    @Published private var addedFields: [CustomFieldUI] = []
+    var hasChanges: Bool {
+        !editedFields.isEmpty || !addedFields.isEmpty
+    }
+
+    init(customFields: [CustomFieldViewModel]) {
+        self.originalCustomFields = customFields
+        updateCombinedList()
+    }
+}
+
+// MARK: - Items actions
+extension CustomFieldsListViewModel {
+    func editField(at index: Int, newField: CustomFieldUI) {
+        guard index >= 0 && index < combinedList.count else {
+            DDLogError("⛔️ Error: Invalid index for editing a custom field")
+            return
+        }
+
+        let oldField = combinedList[index]
+
+        if newField.id == nil {
+            // If there's no id, it means we're now editing a newly added field.
+            if let addedIndex = addedFields.firstIndex(where: { $0.key == oldField.key }) {
+                if newField.key == oldField.key {
+                    // If the key hasn't changed, update the existing added field
+                    addedFields[addedIndex] = newField
+                } else {
+                    // If the key has changed, remove the old field and add the new one
+                    addedFields.remove(at: addedIndex)
+                    addedFields.append(newField)
+                }
+            } else {
+                // This case should not happen in normal flow
+                DDLogError("⛔️ Error: Editing a newly updated field that doesn't exist in combinedList")
+
```

Comment:
nit: empty return that should be deleted

Reply:
Removed as part of refactor https://github.com/woocommerce/woocommerce-ios/pull/14029/commits/98cec53a245a4d41c67c566e7ecab91ad485c229